### ISSUE TITLE
Adapter modules for xfer learning

### ISSUE THREAD

**Issue by erip** [marian-nmt/marian-dev#431]
[This paper](https://arxiv.org/pdf/1902.00751.pdf) looks especially interesting for effectively transferring pre-trained encodings into a downstream task at a fraction of the parameter-update cost. It might be interesting to see how this can be rolled into marian for translation. I think this will practically rely on interop of the pre-trained BERT tf models within marian.

**Comment by emjotde** [comment]
Hi,
Is there an advantage over normal fine-tuning? We do have a BERT model now and we can convert TF models to Marian via a Python script. So, we can just finetune the normal model. The BERT stuff is not documented or announced yet, I still working on it a bit.

**Comment by erip** [comment]
The (truncated) abstract might be helpful, so I've included it below:

>Fine-tuning large pre-trained models is an effective transfer mechanism in NLP. **However, in the presence of many downstream tasks, fine-tuning is parameter inefficient: an entire new model is required for every task.** As an alternative, we propose transfer with adapter modules. Adapter modules yield a compact and extensible model; they add only a few trainable parameters per task, and new tasks can be added without revisiting previous ones. [...] To demonstrate adapter’s effectiveness, we transfer the recently proposed BERT Transformer model [...] On GLUE, we attain within 0.4% of the performance of full fine-tuning, adding only 3.6% parameters per task. By contrast, fine-tuning trains 100% of the parameters per task.

Given my understanding, this would speed up training since there would be significatly fewer updates.

**Comment by emjotde** [comment]
Speed is hardly an issue for finetuning though. This takes minutes for those tiny data sets they use for the down-stream tasks. More likely this is a space-constraint. Let's say an adapter module weighs only 10MB versus a 500MB BERT model or something. 

The forward step with adapters will still be as slow as normal training, you save time in the backwards step, but the gradients still need to be propagated, just not into the BERT parameters. Total compute would barely change, I think? Maybe convergence a bit?

**Comment by erip** [comment]
@emjotde that may be so -- I'd certainly expect faster convergence. There is some sample Python code for the original residual adapters paper on github, so maybe it's worth leaving this open but keeping it low priority. If I get some time, I could try to port it over to C++ and compare.

**Comment by MiroFurtado** [comment]
Chiming in to this still open issue to say that fine-tuning at scale is still a usecase. Or if you need to fine tune for many different applications a BERT language model that has been trained on your particular dataset.

**Comment by emjotde** [comment]
Closing this, feel free to re-open if someone wants to implement this and needs help.
